nibabel/arrayproxy.py

```diff
@@ -53,7 +57,24 @@
 KEEP_FILE_OPEN_DEFAULT = False
 
 
-class ArrayProxy:
+class ArrayLike(ty.Protocol):
+    """Protocol for numpy ndarray-like objects
+
+    This is more stringent than :class:`numpy.typing.ArrayLike`, but guarantees
+    access to shape, ndim and slicing.
+    """
+
+    shape: tuple[int, ...]
+    ndim: int
+
+    def __array__(self, dtype: npt.DTypeLike | None = None, /) -> npt.NDArray:
```

Comment:
I'm sorry if I'm missing something, but shouldn't this be:
```suggestion
    def __array__(self, *, dtype: npt.DTypeLike | None = None) -> npt.NDArray:
```
I think `/` is only used for positional arguments, whereas keyword arguments require `*`. Sorry if I'm missing something.

Reply:
```
numpy.ndarray.__array__ = __array__(...)
    a.__array__([dtype], /) -> reference if type unchanged, copy otherwise.
    
    Returns either a new reference to self if dtype is not given or a new array
    of provided data type if dtype is different from the current dtype of the
    array.
```

Reply:
Huh. Yeah. I am missing something. If you understand why/how and wouldn't mind explaining sometime, I'd be happy to learn.

Reply:
I thought that once a default value is assigned (including `None`, as it is here), the parameter is immediately considered a keyword argument, and that `/` is used only for positional arguments. 
Looking at [the relevant PEP](https://peps.python.org/pep-0570/#syntax-and-semantics), I learned that: 

> The logic around optional values for positional-only parameters remains the same as for positional-or-keyword parameters.